cmd/gcp-controller-manager/istiod_csr_approver.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"context"
+	"strings"
+
+	capi "k8s.io/api/certificates/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	certutil "k8s.io/kubernetes/pkg/apis/certificates/v1"
+	"k8s.io/kubernetes/pkg/controller/certificates"
+)
+
+const istiodSignerName = "pki.gke.io/istiod"
+
+func newIstiodApprover(ctx *controllerContext) *istiodApprover {
+	return &istiodApprover{
+		ctx: ctx,
+	}
+}
+
+type istiodApprover struct {
+	ctx *controllerContext
+}
+
+func (a *istiodApprover) handle(csr *capi.CertificateSigningRequest) error {
+	if csr.Spec.SignerName != istiodSignerName {
+		return nil
+	}
+	if approved, denied := certificates.GetCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+
+	if !hasExactUsages(csr, []capi.KeyUsage{
+		capi.UsageKeyEncipherment,
+		capi.UsageDigitalSignature,
+		capi.UsageServerAuth,
+	}) {
+		return a.deny(csr, "disallowed usages requested")
+	}
+
+	x509cr, err := certutil.ParseCSR(csr.Spec.Request)
+	if err != nil {
+		return a.deny(csr, "unable to parse csr")
+	}
+
+	if len(x509cr.URIs) != 0 || len(x509cr.EmailAddresses) != 0 || len(x509cr.IPAddresses) != 0 {
+		return a.deny(csr, "disallowed sans requested")
+	}
+
+	if x509cr.Subject.CommonName != csr.Spec.Username {
```

Comment:
could we also check if CommonName exists and then do a comparison?

Reply:
Done.